### ISSUE TITLE
Smilei crash at random timestep 

### ISSUE THREAD

**Issue by robintimmis** [SmileiPIC/Smilei#578]
## Description

For the past couple months I have been experiencing random crashes of my simulations. Random in the sense that the timestep of the crash is not the same. In the crash the output file stops writing, but the simulation keeps going. It is like the simulation is unable to advance but it is still trying to. The simulation continues running until the time allocated for the job runs out (leading to very expensive waste of resources for large jobs). I have been liaising with the ARCHER2 support team, so far they have not been able to identify the issue.

No error messages since the simulation seems not to know that there is an issue.

## Steps to reproduce the problem

It seems completely arbitrary, simply the larger the job the more likely it is to happen, probably because there is more opportunity for it to happen. Sometimes it happens after a few timesteps, sometimes tens of thousands of timesteps.

## Parameters

Not sure how to find the Smilei version but I recompiled it 2 weeks ago from github.

Input file (namelist) and script to run on ARCHER2
[submission_namelist.zip](https://github.com/SmileiPIC/Smilei/files/10085185/submission_namelist.zip)

Output file of simulation which failed to run
[output_file.zip](https://github.com/SmileiPIC/Smilei/files/10085193/output_file.zip)

Compilation commands for compiling smilei
[compile.txt](https://github.com/SmileiPIC/Smilei/files/10085231/compile.txt)

Details on archer
https://www.archer2.ac.uk/about/hardware.html

  - C++ compiler (to get some information `g++ --version`) - g++ (GCC) 11.2.0 20210728 (Cray Inc.)
  - MPI version (`mpic++ --version`, `mpic++ -show`) (not found)
  - HDF5 version (`h5cc --version`) - gcc (GCC) 11.2.0 20210728 (Cray Inc.)
  - Python version (`python --version`) - Python 3.9.7

**Comment by beck-llr** [comment]
Hi
Thanks for all the info, it's always nice to exchange with users.

There are 8 NUMA regions on your compute node so my guess would have been to use 8 tasks per node instead of 64 and 16 openMP threads instead of 2. But I guess you've tried different configurations and found out this was better ?
Also we strongly advise to use
`export OMP_SCHEDULE=dynamic`

What you're saying is that you are experiencing random deadlocks. The code just stops running without crashing. This is usually the consequence of a faulty MPI communication. It would be nice to know which MPI distribution you are using. 

If you're lucky, reducing the number of MPI processes and increasing the number of openMP threads might just do the trick for you. There could be an MPI message tagging issue with that many MPI processes (that depends on the MPI distribution).

I can also comment on your number of patches. You have exactly as many patches as threads (8192). We strongly advise to have strictly more patches than threads (it is even in comment in your namelist). Since your patches are already quite small, you may get similar performance using less compute nodes. This could potentially also help solving the faulty MPI communication issue.  Your simulation just looks "too small" for that many compute cores. 

To sum up, my recommendation is to try running the same namelist with 32 or 16 nodes,  8 tasks per node and 16 openMP threads.  Or at least, try to go in the direction of reducing your total number of MPI processes. Let me know how that goes.

Apart from that, I see that you are using  radiation reaction and Breit Wheeler which I am not an expert of so I'll leave further comments about these to other people.

**Comment by mpikzink** [comment]
Hi @all,

we do have the same issue on our HPC setup:
```
gcc (GCC) 9.2.0
Python 3.9.6
g++ (GCC) 9.2.0
hdf5-1.12.2
openmpi-4.1.4
```
Hardware is almost the the same:

Nodes = 4 nodes
Processor | 2× AMD EPYC 7742, 2.25 GHz, 64-core
Cores per node | 128 (2× 64-core processors)
NUMA structure | No numa activated (1 region)
Memory per node | 1024 GiB 
Memory per core = 8GB
Interconnect = 1 × 25 Gbps bi-directional per node Ethernet

We launched our simulation twice using two different slurmjobs.  
In the first run we requested 4 nodes (exclusive)  using only 1 openMP thread and everything ended fine.
In the second simulation, we still requested 4 nodes but increased the number of openMP threads to 16 (using 32 MPI tasks): the simulation got stuck with the same behavior as that described by @robintimmis. Indeed, we tried several times only to see that the timestep where the simulation freezes is always different.

**Comment by beck-llr** [comment]
Hi
Could you please provide a namelist ? Do you also use Radiation Reaction and/or Breit wheeler ?
Thanks

**Comment by mccoys** [comment]
Also please try to recompile with `config=no_mpi_tm`

**Comment by mpikzink** [comment]
[beam_fields_reconstruction_Au.py.txt](https://github.com/SmileiPIC/Smilei/files/10095077/beam_fields_reconstruction_Au.py.txt)

**Comment by xxirii** [comment]
Dear @robintimmis and @mpikzink,

I have started to study your cases.

First, I will explain you why @mccoys is suggesting to compile with the flag `config=no_mpi_tm`. In the past, we have experienced some similar issues coming from MPI. Some MPI implementation supports badly the default behavior in Smilei (so called threadmultiple) or they are not compiled with the proper options. Sometime, Smilei is unable to detect that this feature is not available. The consequence is very similar to what you observe. The code get stuck at some random iteration without any error message. One way to avoid this is to compile with `config=no_mpi_tm`. We do have experienced that with openmpi not compiled correctly.

Then, for both of you, I miss the error output. You only sent me the standard output. The error output (*.err maybe) contains possible error messages that may guide us.

**Comment by xxirii** [comment]
@mpikzink can you provide a namelist without the external particles that will certainly crash on your system:

```
xyzw_electron_beam = np.load('electron_beam_truncated_xyzw_after_6foil_modified.npz')['data']
pxyz_electron_beam = np.load('electron_beam_truncated_pxyz_after_6foil_times_10.npz')['data']
```

I am currently testing on a machine equipped with similar hardware (TGCC Irene Joliot-curie - AMD Rome) but it is not a Cray machine so the environment that you have cannot be reproduced exactly.

@robintimmis the comment of @beck-llr is important, I think you use too much resources for your run by having a number of patchs equal to the number of cores. You have a lot of very small mpi domains. I would recommend to use less resources in fact. I am still testing your case and will let you know.

**Comment by beck-llr** [comment]
> Then, for both of you, I miss the error output. You only sent me the standard output. The error output (*.err maybe) contains possible error messages that may guide us.

Unfortunately, I don't think there is any error message since the code just falls in a deadlock and doesn't do anything anymore. At least that is what I understood from the issue description and I know it can happen.

**Comment by xxirii** [comment]
Yes, but I want it to be sure.

**Comment by xxirii** [comment]
Hi @ robintimmis ,

I have tested extensively your case and was not able to reproduce your bug.

At first, I was having a out-of-memory error always at timestep 9625. I finally realized that your diag ParticleBinning was the cause. 

```python
DiagParticleBinning(
    name = "electron_4D_phase_space",
    deposited_quantity = "weight",
    every = [9625,10000],
    time_average = 1,
    species = ["electron"],
    axes = [
        ["x", 14.7, 16.6, 100],
        ["px", 200, 0., 200],
        ["y", 2.64, 5.78, 100],
        ["py", -200, 0., 200]
    ]
)
```

Your 4d diag is quite heavy. To compute it, Smilei will initialize per MPI domain a grid of the corresponding diag size (200*200*100*100 roughly 3 Gb). In you case, with 64 MPI processes per node, it explodes the total RAM memory available per node. 
Then, we have discovered an error in the definition of the same diag, for `px` the max value is less than the min value.

If I comment this diag, you configurations runs perfectly well for me using 64 MPIs per node and 2 threads per MPI. Moreover, I am using just 8 nodes instead of 64 and the job finishes in 40 minutes which is not that much (roughly 2000 core hours).

Finally, your issue looks like an MPI issue maybe related to what we mentioned before. What I suggest:
1) correct the diag
2) recompile with the config=`no_mpi_tm` mode (see the doc for more info) : 
```
make machine=archer2 -j 8 config=`no_mpi_tm`
```
3) run you case using less MPI and more OpenMP threads as suggested by @beck-llr 

If you still have problems, we will continue to explore (maybe try to access archer to test).

Thank you.
